crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -75,46 +76,63 @@ impl Fulfillment {
+                    *max_volume_factor,
+                ),
+                FeePolicy::PriceImprovement {
+                    factor,
+                    max_volume_factor,
+                    quote,
+                } => {
+                    let price_limits = adjust_quote_to_order_limits(
+                        Order {
+                            sell_amount: current_order_price_limits.sell.0,
+                            buy_amount: current_order_price_limits.buy.0,
+                            side: self.order().side,
+                        },
+                        Quote {
+                            sell_amount: quote.sell.amount.0,
+                            buy_amount: quote.buy.amount.0,
+                            fee_amount: quote.fee.amount.0,
+                        },
+                    )?;
+                    self.calculate_fee(price_limits, prices, *factor, *max_volume_factor)
+                }
+                FeePolicy::Volume { factor } => self.fee_from_volume(prices, *factor),
+            }?;
 
-        match self.order().protocol_fees.first() {
-            Some(FeePolicy::Surplus {
-                factor,
-                max_volume_factor,
-            }) => self.calculate_fee(
-                PriceLimits {
-                    sell: self.order().sell.amount,
-                    buy: self.order().buy.amount,
-                },
-                prices,
-                *factor,
-                *max_volume_factor,
-            ),
-            Some(FeePolicy::PriceImprovement {
-                factor,
-                max_volume_factor,
-                quote,
-            }) => {
-                let price_limits = adjust_quote_to_order_limits(
-                    Order {
-                        sell_amount: self.order().sell.amount.0,
-                        buy_amount: self.order().buy.amount.0,
-                        side: self.order().side,
-                    },
-                    Quote {
-                        sell_amount: quote.sell.amount.0,
-                        buy_amount: quote.buy.amount.0,
-                        fee_amount: quote.fee.amount.0,
-                    },
-                )?;
-                self.calculate_fee(price_limits, prices, *factor, *max_volume_factor)
+            // Recalculate the current price limits in order to apply the next protocol fees
+            // to the new amount
+            match self.order().side {
+                Side::Buy => {
+                    current_order_price_limits.buy = current_order_price_limits
+                        .buy
+                        .checked_add(&current_fee)
+                        .ok_or(Math::Overflow)?;
+                }
+                Side::Sell => {
+                    current_order_price_limits.sell = current_order_price_limits
+                        .sell
+                        .checked_sub(&current_fee)
+                        .ok_or(Math::Negative)?;
+                }
```

Comment:
Why recalculating the price limits? Why not calculating the protocol fee for first policy, then adjusting `executed` and `fee` of the `Fulfilment`, and then calculating next fee policy etc? 

I find it very hard to understand the logic behind altering the limit prices, even if it works somehow (but I might be biased since I already know this code well).